tensornetwork/backends/jax/gmres.py

```diff
@@ -0,0 +1,225 @@
+
+def gmres_m(A_mv: Callable, A_args: Sequence, 
+            b: jnp.ShapedArray, x0: jnp.ShapedArray, tol: float, atol: float,
+            num_krylov_vectors: int,
+            maxiter: int) -> Tuple[jnp.ShapedArray, float, int, bool]:
+  """
+  Solve A x = b for x using the m-restarted GMRES method. This is
+  intended to be called via jax_backend.gmres.
+
+  Given a linear mapping with (n x n) matrix representation
+      A = A_mv(*A_args) gmres_m solves
+      Ax = b          (1)
+  where x and b are length-b vectors, using the method of
+  Generalized Minimum RESiduals with M iterations per restart (GMRES_M).
+
+  Args:
+
+  A_mv     : A function `v0 = A_mv(v, *A_args, **A_kwargs)` where `v0` and
+             `v` have the same shape.
+  b        : The `b` in `A @ x = b`.
+  A_args   : Positional arguments to `A_mv`.
+  x0       : Initial guess solution.
+  tol, atol: Solution tolerance to achieve,
+             norm(residual) <= max(tol*norm(b), atol).
+  num_krylov_vectors
+           : Size of the Krylov space to build at each restart.
+  maxiter  : The Krylov space will be repeatedly rebuilt up to this many
+             times.
+
+
+  RETURNS
+  -------
+  x (array, (n,)) : The approximate solution.
+  beta (float)    : Norm of the residual at termination.
+  n_iter (int)    : Number of iterations at termination.
+  converged (bool): Whether the desired tolerance was achieved.
+  """
+  num_krylov_vectors += 1
+  x = x0
+  converged = False
+  r, beta = gmres_residual(A_mv, A_args, b, x)
+  b_norm = jnp.linalg.norm(b)
+  for n_iter in range(maxiter):
+    # pylint: disable=too-many-function-args
+    x = gmres(A_mv, A_args, num_krylov_vectors, x, r, beta)
+    r, beta = gmres_residual(A_mv, A_args, b, x)
+    if beta <= max(tol*b_norm, atol):
+      converged = True
+      break
+  return (x, beta, n_iter, converged)
+
+
+@jax.jit
+def gmres_residual(A_mv: Callable, A_args: Sequence, b: jnp.ShapedArray,
+                   x: jnp.ShapedArray) -> Tuple[jnp.ShapedArray, float]:
+  """
+  Computes the residual vector r and its norm, beta, which is minimized by
+  GMRES.
+  """
+  r = b - A_mv(x, *A_args)
+  beta = jnp.linalg.norm(r)
+  return r, beta
+
+
+@partial(jax.jit, static_argnums=(2,))
+def gmres(A_mv: Callable, A_args: Sequence, n_kry: int,
+          x0: jnp.ShapedArray, r: jnp.ShapedArray,
+          beta: float) -> jnp.ShapedArray:
+  """
+  Solve A x = b for x by the unrestarted GMRES method.
+  Given A, a trial solution x, the residual r,
+  and the size n_kry of the Krylov space, iterates x towards the solution,
+  by finding y in x = x_0 + V y minimizing ||beta - H y||.
+  """
+  v = r / beta
+  Vk_1, Htilde = gmres_arnoldi(A_mv, A_args, n_kry, v)
+  Q, Rtilde = jnp.linalg.qr(Htilde, mode="complete")
+  Q = Q.T.conj()
+  R = Rtilde[:-1, :]
+  g = beta*jnp.ravel(Q[:-1, 0])
+  y = jax.scipy.linalg.solve_triangular(R, g)
+  update = Vk_1[:, :-1] @ y
+  x = x0 + update
+  return x
+
+
+@partial(jax.jit, static_argnums=(2,))
```

Comment:
Is this returning an arnoldi factorization?

Reply:
If so, do you think you can use the one that is implemented in the backend?

Reply:
Well, the one in the backend returns the eigendecomposition; I need the V and H from the Krylov decomposition (actually I need V and the full QR decomposition of H). If Arnoldi were refactored to produce these it could be used.

However, a more efficient implementation, which I will add later, would also update the GMRES residual by a Givens rotation once per new Krylov vector. This becomes specialized enough that I think it makes more sense to keep things together with GMRES until proved otherwise, which is reflected in the code as written.

Reply:
sorry I was not precise in my wording. I meant the arnoldi factorization in jitted_functions.py. This one priduces exactly that

Reply:
would it make sense to move gmres into the jitted_functions.py file?

Reply:
My recent push should address the above